core/src/main/kotlin/com/bumble/appyx/core/children/ChildNodeCreationManager.kt

```diff
@@ -15,88 +16,141 @@ import kotlinx.coroutines.flow.update
 import kotlinx.coroutines.flow.updateAndGet
 import kotlinx.coroutines.launch
 
+/**
+ * Initializes and removes nodes based on parent node routing source.
+ *
+ * Lifecycle of these nodes is managed in [com.bumble.appyx.core.lifecycle.ChildNodeLifecycleManager].
+ */
 internal class ChildNodeCreationManager<Routing : Any>(
     private var savedStateMap: SavedStateMap?,
     private val customisations: NodeCustomisationDirectory,
-    private val childMode: ChildEntry.ChildMode,
+    private val keepMode: ChildEntry.KeepMode,
 ) {
     private val _children =
         MutableStateFlow<Map<RoutingKey<Routing>, ChildEntry<Routing>>>(emptyMap())
     val children: StateFlow<ChildEntryMap<Routing>> = _children.asStateFlow()
+    private lateinit var parentNode: ParentNode<Routing>
 
     fun launch(parentNode: ParentNode<Routing>) {
-        savedStateMap.restoreChildren(parentNode)?.also { restoredMap ->
+        this.parentNode = parentNode
+        savedStateMap.restoreChildren()?.also { restoredMap ->
             _children.update { restoredMap }
             savedStateMap = null
         }
+        syncNavModelWithChildren(parentNode)
+    }
+
+    private fun syncNavModelWithChildren(parentNode: ParentNode<Routing>) {
         parentNode.lifecycle.coroutineScope.launch {
-            parentNode.navModel.elements.collect { elements ->
-                _children.update { map ->
-                    val navModelKeys = elements
-                        .mapTo(HashSet(elements.size, 1f)) { element -> element.key }
-                    val localKeys = map.keys
-                    val newKeys = navModelKeys - localKeys
-                    val removedKeys = localKeys - navModelKeys
-                    val mutableMap = map.toMutableMap()
-                    newKeys.forEach { key ->
-                        mutableMap[key] =
-                            ChildEntry.create(
-                                key = key,
-                                resolver = parentNode,
-                                buildContext = null.toBuildContext(parentNode),
-                                childMode = childMode,
-                            )
+            parentNode.navModel.screenState.collect { state ->
+                val navModelOnScreenKeys: Set<RoutingKey<Routing>>
+                val navModelOffScreenKeys: Set<RoutingKey<Routing>>
+                val navModelKeys: Set<RoutingKey<Routing>>
+                when (keepMode) {
+                    ChildEntry.KeepMode.KEEP -> {
+                        // Consider everything as on-screen for keep mode
+                        navModelOnScreenKeys =
```

Comment:
Perhaps this variable name is incorrect if we treat offScreen as onScreen? How about `navModelAvailableKeys` or `navModelKeptKeys`? (kept being the past tense of keep)

Reply:
If we change this, we should also change `updateChildren`